examples/with-apollo/pages/ssr.js

```diff
@@ -0,0 +1,33 @@
+import App from '../components/App'
+import InfoBox from '../components/InfoBox'
+import Header from '../components/Header'
+import Submit from '../components/Submit'
+import PostList, {
+  ALL_POSTS_QUERY,
+  allPostsQueryVars,
+} from '../components/PostList'
+import { initializeApollo, addApolloState } from '../lib/apolloClient'
+
+const SSRPage = () => (
+  <App>
+    <Header />
+    <InfoBox>ℹ️ This page shows how to use SSR with Apollo.</InfoBox>
+    <Submit />
+    <PostList />
+  </App>
+)
+
+export async function getServerSideProps() {
+  const apolloClient = initializeApollo()
+
+  await apolloClient.query({
+    query: ALL_POSTS_QUERY,
+    variables: allPostsQueryVars,
+  })
+
+  return addApolloState(apolloClient, {
+    props: {},
```

Comment:
Wouldn't you want to render the posts and use the data you fetched?

Reply:
Here, we are returning `apolloClient`, which contains fetched data, and it adds all the posts to the apollo cache, we have at a client.

Then, inside the `<PostList />` component, that data is queried and the apollo client provides it from the cache.

We are using the same technique in the SSG example to use fetched data.

```js
return addApollState(apolloClient, {
   props: {},
}
```

Reply:
Intersting. I know that's not your fault, because you're just following the pattern already set, but that seems confusing. All other usage of `getStaticProps` / `getServerSideProps` show forwarded the data to the component via props (hence the names). I wonder how many people are getting stuck on this.

Reply:
Okay. So, for this example should we follow another pattern, which is passing the posts via props and render them?

Reply:
@leerob I see we are using pagination in `PostList` component, hence we need `useQuery` hook to use `fetchMore` function.
For this example, if we pass posts props via `getStaticProps` / `getServerSideProps` and render them, we will not be able to use `fetchMore`. Any suggestions?

Reply:
> Interesting. I know that's not your responsibility, because you're just following the pattern already set, but that seems confusing. All other usage of `getStaticProps` / `getServerSideProps` show forwarded the data to the component via props (hence the names). I wonder how many people are getting stuck on this.

I've adopted the same convention as above following the examples available, but if there are some better practice / more straightforward I'd love to hear about them!

Reply:
We ran into this issue/decision on our project. Why isn't the best practice to just create an `ApolloClient` inside `gSSP`, fetch whatever and pass it as normal props to the page?

Reply:
I think it is because, if we have to use that data in some other component, you can access it via the apollo client cache. Also, in the current component, if we intend to use functionality like "fetchMore" like in this example, this is a considerable practice to fetch data at the server and return apollo client at server-side to merge that data with apollo client on the client-side. Just by passing data as props, you can not go any further with that data except rendering. Is this making sense?

Reply:
Yes. What seems weird to me is that everything you described feels more like client-side rendering behaviour (like wanting different pages to share Apollo cache and fetch more). If all you want is client side hydration on the first render, why not just pass normal data and manually merge it with the Apollo cache once you get to the client?

Reply:
It seems like you're saving one or two lines of code that belong to app land at the cost of abstracting away all the merging logic, etc, when this probably should be done on a page-by-page basis. Does this make sense to you?

I understand what you are trying to do, and it makes sense from a certain point of view, but it seems to conflict with Next's philosophy of clearly separated pages and client/server side stuff. Feels like it's blurring the line, to me at least

Reply:
@filipemarins You made a point. It makes sense. Okay so let's forget about Apollo Cache, and consider we are fetching data and passing it via props. Is there any way to implement pagination in this example we are using "fetchMore" from Apollo Client to make it work. Can we pass "fetchMore" as props? And call it whenever we need the next data?

Reply:
That's the issue: you can't because functions are not JSONable, and what Next abstracts from us is basically a REST call (or something). But you don't have to right? `fetchMore` is only sugar for passing just a cursor (and Apollo will fill the rest of the vars with whatever was there before). If you hydrate on the first render on the client side, you should still be perfectly able to fetch the cursor from your fetched data and pass it to `fetchMore`. Really the only difference is that the first render will have cached Apollo Client data. This is the only connecting between Apollo and Next (with SSR). You fill the cache with whatever arbitrary data you want (most of the times, the first data the client will need). From the pages side, none of this matters, they just do whatever queries they need, and Apollo will have them in cache, they code stays the same. The only issue here (I think) is how to hydrate this state, and IMO it can be done with `useApolloClient` (get the client, hydrate data, etc etc), right before you `useQuery`. Of course you can abstract this away with a custom hook to reuse between pages. Something like `useHydratedQuery`.

Reply:
I see that makes sense. You are right, making a custom hook for this purpose would be really beneficial. But adding such complex logic inside `with-apollo`, wouldn't that be overkill for people who are just getting started. I think the current implementation is good, from the perspective of the example. We ain't creating a full-fledged app in here. Haha.

Reply:
Yeah that's also true. I mean if it works, it can't be that bad. But note that this current example (and the ones that existed before) doesn't point (IMO) in the best direction, which basically would be align with Next's way of thinking. With this said, it's not _that_ complex, and I think anyone who is serious about Next will ask the same question we are asking and decide for themselves.

As for the hook, idk, it's as complex as what we currently have. Tbh I don't have a strong opinion to leave or remove, I'm just curious to see what core thinks (@leerob ?)

Reply:
Yep same. Let's see.

Reply:
I agree that both approaches work. I just worry about beginners seeing it and getting confused.

Maybe it just needs a comment to explain why?